simvue/eco/emissions_monitor.py

```diff
@@ -0,0 +1,325 @@
+"""
+CO2 Monitor
+===========
+
+Provides an interface for estimating CO2 usage for processes on the CPU.
+"""
+
+__author__ = "Kristian Zarebski"
+__version__ = "0.1.0"
+__license__ = "MIT"
+__date__ = "2025-02-27"
+
+import datetime
+import json
+import pydantic
+import dataclasses
+import threading
+import time
+import logging
+import typing
+import psutil
+import humanfriendly
+import pathlib
+import os
+
+from simvue.eco.api_client import APIClient, CO2SignalResponse
+
+TIME_FORMAT: str = "%Y_%m_%d_%H_%M_%S"
+
+
+@dataclasses.dataclass
+class ProcessData:
+    process: psutil.Process
+    cpu_percentage: float = 0.0
+    power_usage: float = 0.0
+    total_energy: float = 0.0
+    energy_delta: float = 0.0
+    co2_emission: float = 0.0
+    co2_delta: float = 0.0
+
+
+class CO2Monitor(pydantic.BaseModel):
+    """
+    CO2 Monitor
+
+    Provides an interface for estimating CO2 usage for processes on the CPU.
+    """
+
+    thermal_design_power_per_core: pydantic.PositiveFloat | None
+    cpu_idle_power: pydantic.PositiveFloat
+    cpu_interval: float = 1.0
+    local_data_directory: pydantic.DirectoryPath
+    intensity_refresh_rate: int | None | str
+    co2_intensity: float | None
+    co2_signal_api_token: str | None
+
+    def now(self) -> str:
+        """Return data file timestamp for the current time"""
+        _now: datetime.datetime = datetime.datetime.now(datetime.UTC)
+        return _now.strftime(TIME_FORMAT)
+
+    @property
+    def outdated(self) -> bool:
+        """Checks if the current data is out of date."""
+        if not self.intensity_refresh_rate:
+            return False
+
+        _now: datetime.datetime = datetime.datetime.now()
+        _last_updated: str = self._local_data["last_updated"]
+        _latest_time: datetime.datetime = datetime.datetime.strptime(
+            _last_updated, TIME_FORMAT
+        )
+        return (_now - _latest_time).seconds < self.intensity_refresh_rate
+
+    def _load_local_data(self) -> dict[str, str | dict[str, str | float]] | None:
+        """Loads locally stored CO2 intensity data"""
+        self._data_file_path = self.local_data_directory.joinpath(
+            "ecoclient_co2_intensity.json"
+        )
+
+        if not self._data_file_path.exists():
+            return None
+
+        with self._data_file_path.open() as in_f:
+            _data: dict[str, str | dict[str, str | float]] | None = json.load(in_f)
+
+        return _data or None
+
+    def __init__(self, *args, **kwargs) -> None:
+        """Initialise a CO2 Monitor.
+
+        Parameters
+        ----------
+        thermal_design_power_per_core: float | None
+            the TDP value for the CPU. Default of None uses naive 85W value.

```

Comment:
This is an important assumption and should be listed in the doc with a reference here. Desktops typically have 65W (e.g. i5) so its sensible that its slightly higher unless specified.

Reply:
Added to https://github.com/simvue-io/docs/pull/61